File: csharp/Day07.cs
internal class Day07 : IDay
{
    public int DayNumber => 7;
    public int PartOneTestAnswer => 37;
    public int PartTwoTestAnswer => 168;

    public int ExecutePartOne(string[] input)
    {
        return CalculateFuelSpend(input, true);
    }

    public int ExecutePartTwo(string[] input)
    {
        return CalculateFuelSpend(input, false);
    }

    private static int CalculateFuelSpend(string[] input, bool isConstantRate)
    {
        var numbers = input.First().Split(',').Select(int.Parse).ToList();
        var bestNumberSum = int.MaxValue;
        for (var number = numbers.Min(); number < numbers.Max(); number++)
        {
            var numberSum = numbers.Sum(x =>
            {
                var horizontalDistance = Math.Abs(x - number);
                if (isConstantRate)
                {
                    return horizontalDistance;
                }
                return Enumerable.Range(1, horizontalDistance).Sum();
            });
            if (numberSum < bestNumberSum)
            {
                bestNumberSum = numberSum;
            }
        }
        return bestNumberSum;
    }
}

File: typescript/2024/day07.ts
import { Day } from "../program";

export class Day07 implements Day {
  partOneTestAnswer = 3749;
  partTwoTestAnswer = 11387;

  executePartOne(input: string[]): number {
    return input
      .filter((calibrationEquation) =>
        this.canBeValid(calibrationEquation, false),
      )
      .map((calibrationEquation) => Number(calibrationEquation.split(":")[0]))
      .reduce((a, b) => a + b);
  }

  executePartTwo(input: string[]): number {
    return input
      .filter((calibrationEquation) =>
        this.canBeValid(calibrationEquation, true),
      )
      .map((calibrationEquation) => Number(calibrationEquation.split(":")[0]))
      .reduce((a, b) => a + b);
  }

  canBeValid(input: string, concatOperatorAllowed: boolean) {
    const numberOfOperators = concatOperatorAllowed ? 3 : 2;
    const [expectedSum, allNumbers] = input.split(": ");
    const numbers = allNumbers.split(" ").map(Number);
    const amountOfTries = Math.pow(numberOfOperators, numbers.length - 1);
    for (let i = 0; i < amountOfTries; i++) {
      let currentTotal = numbers[0];
      for (let j = 1; j < numbers.length; j++) {
        switch (
          Math.floor(i / Math.pow(numberOfOperators, j - 1)) % numberOfOperators
        ) {
          case 0:
            currentTotal = currentTotal + numbers[j];
            break;
          case 1:
            currentTotal = currentTotal * numbers[j];
            break;
          case 2:
            currentTotal = Number(`${currentTotal}${numbers[j]}`);
            break;
        }
      }
      if (currentTotal === Number(expectedSum)) {
        return true;
      }
    }
    return false;
  }
}
